### DISCUSSION TITLE
Latest version of Copilot on Neovim doesn't support latest version on Node.js

### DISCUSSION THREAD

**Discussion by mystilleef** [community/community#16800] [Copilot Conversations] [answered]
I updated the `Neovim` version of `Copilot` to `1.3.1`. And now it claims that the current version of `Node.js` `18.2.0` is no longer supported. That's odd because `1.3.0` and earlier versions worked perfectly fine with the latest version of Node.js. Switching `Node.js` to the `LTS` release just to use `Copilot` is a huge inconvenience, as most of my Node projects and dependencies already use the `current` version of `Node.js`.

The temporary workaround for those who are interested is to pin `Copilot` on `Neovim` to `1.3.0`.

**Comment by tpope** [comment, marked as answer]
**Edit:** Node.js 18+ is supported now.

---

The issue only manifests on some source languages: JavaScript, TypeScript, Python, Ruby, and Go. Downgrading Copilot.vim is a temporarily workaround; support for it will be eventually dropped.

If you'd like to have multiple versions of Node.js installed, you can use a tool like [nodenv](https://github.com/nodenv/nodenv) or [nvm](https://github.com/nvm-sh/nvm) to install them in your home directory, and then tell Copilot where to find the appropriate version:

```vim
 let g:copilot_node_command = "~/.nodenv/versions/16.15.0/bin/node"
```

**Reply by mystilleef** [reply]
Thanks, that works. 

For `nvm` users, here's the default location for the `LTS` release.

```lua
-- lua
vim.g.copilot_node_command = "~/.nvm/versions/node/v16.15.0/bin/node"
```

```vim
" vimscript
let g:copilot_node_command = "~/.nvm/versions/node/v16.15.0/bin/node"
```

**Reply by alehatsman** [reply]
thanks @tpope, solved the problem

**Reply by AlejandroFNadal** [reply]
thanks to both of you @tpope and @mystilleef , that worked very well. If you want to use the latest working node version for the plugin, you can run `nvm ls-remote 17.*` and you will see the list of the minor versions inside 17.

**Comment by ouuan** [comment]
Any plan to support Node 18? I don't know the exact reason why copilot.vim does not support Node 18, but it doesn't seem like a good idea to always stick to older versions of Node. There have been a few months and a few releases of copilot.vim since this issue, so I'm wondering if the support of Node 18 will be added in the near future.

**Reply by tpope** [reply]
There's a bug in an upstream dependency. It's a popular dependency so it's not going to be forgotten. Node 16 is still the default recommendation at <https://nodejs.org/en/download/>; we have a while before it's considered an "older version".

**Reply by vegerot** [reply]
Which upstream dependency?  I'll consider patching that dependency if you would please give me some more details 🙂

**Reply by tpope** [reply]
The issue is <https://github.com/tree-sitter/tree-sitter/issues/1765>. I'm sure contributions would be appreciated. Please everybody practice good etiquette and avoid low value "any update on this?" posts.

**Reply by vegerot** [reply]
Don't worry, it's not Hacktoberfest anymore I'm a real developer 😉

**Reply by thdxr** [reply]
It looks like that upstream dependency is fixed, does that unblock this issue?

EDIT: nevermind didn't realize it wasn't released

**Comment by AlejandroFNadal** [comment]
We are still having this issue, Node.js version 12.x-17.x required but found 19.2

**Comment by fang2hou** [comment]
Here is a nvim lua snippets for people who using volta as node version manager.
You can also use `NODE_16_BIN` env to manually set the node binary externally.

```lua
do
    local node_path = os.getenv("NODE_16_BIN")

    if not node_path then
        local handle = io.popen('echo "console.log(process.argv[0])" | volta run node --node 16', 'r')
        if handle then
            local volta_node_path = handle:read("*l")
            handle:close()

            if volta_node_path and volta_node_path ~= '' then
                node_path = volta_node_path
            end
        end
    end

    if node_path then
        vim.cmd(string.format("let g:copilot_node_command = '%s'", node_path))
    end
end
```

**Comment by bunyk** [comment]
I have no suggestions from copilot.vim 1.13.0, and when using approach suggested here I get "Warning: Ignoring g:copilot_node_command: Node.js 16.15.0 is end-of-life", and it still uses newer Node version.

**Reply by bunyk** [reply]
Degrading to 1.3.0 also does not help because then server refuses communicate with plugin.

**Reply by tpope** [reply]
Node.js 18+ is supported now. I've edited the marked answer to indicate that.
